### ISSUE TITLE
[PT-04] GRPO Group Relative Policy Optimization (Deepseek-R1 livrable cle)

### ISSUE THREAD

**Issue by jsboige** [jsboige/CoursIA#1757]
## Parent
Epic #1742 (RL/Post-training SOTA — Deepseek-Qwen style)
Epic #1454 (Training & Post-Training)

## Scope — LIVRABLE CLE Epic #1742

Notebook `MyIA.AI.Notebooks/PostTraining/PT_04_grpo_deepseek_r1.ipynb` : implementer **Group Relative Policy Optimization** (Shao et al., DeepSeekMath 2024, popularise par Deepseek-R1 janvier 2025).

GRPO = la technique cle de Deepseek-R1. Reduction memoire massive vs PPO classique : pas de critic network (suppression value head), advantage calcule comme normalization au sein d'un **group** d'echantillons pour le meme prompt.

Formule cle (advantage GRPO) :

```
A_i = (R_i - mean({R_1, ..., R_G})) / std({R_1, ..., R_G})
```

ou G = group size (typique 4-16 par prompt), `R_i` = reward (verifiable ou RM). Suppression critic network = -50% memoire training vs PPO.

## Implementation

- Library : `trl.GRPOTrainer` (HF TRL integration 2025)
- Base model : Qwen2.5-0.5B-Instruct (ou SFT/DPO checkpoint si chaine sequentielle)
- Group size : 8 (compromis variance/memoire)
- Reward function : initialement reward function simple (longueur cible, format JSON valide, contient mot-cle) — preparation a PT-05 RLVR avec verifier exact
- KL regularization : beta=0.04 (defaut Deepseek)
- LoRA : r=8, alpha=16, attention+MLP

## Cellules attendues

1. Histoire : pourquoi Deepseek-R1 a popularise GRPO (janvier 2025, reasoning emergent sans SFT)
2. Math GRPO vs PPO : suppression critic, advantage relatif intra-group, intuition variance reduction
3. Pseudocode GRPO step (sample G outputs, compute rewards, advantage normalization, policy update)
4. Setup TRL GRPOTrainer + GRPOConfig
5. Reward function design (callable retournant float par output)
6. Group size hyperparameter exploration (variance vs memoire)
7. Conditional training avec LOAD_MODEL_AND_TRAIN pattern
8. Eval qualitative : exemples outputs avant/apres GRPO sur prompts reasoning
9. 5 pieges GRPO : reward hacking, KL collapse, group size mal calibre, mode collapse, reward sparsity
10. Transition vers PT-05 RLVR (reward verifiable exact pour math/code)

## Acceptance criteria

- LIVRABLE CLE Epic #1742 — qualite irreprochable
- C.1 / C.2 / C.3 strict
- Eval qualitative documentee (memes prompts before/after)
- Pas de claim BEATS quantitative obligatoire (terrain pedagogique) mais documenter behavior change observable
- References : DeepSeekMath paper + Deepseek-R1 paper + TRL GRPOTrainer docs

## Contraintes RTX 3070 8 Go

- 4-bit quantization obligatoire
- Group size 4-8 (memoire scale lineaire avec G)
- max_seq_length <= 512 (reasoning prompts courts)
- gradient checkpointing active
- training time cible : 60-120 min

## References

- Shao et al., "DeepSeekMath: Pushing the Limits of Mathematical Reasoning in Open Language Models" (2024) — paper GRPO origin
- Deepseek-AI, "DeepSeek-R1: Incentivizing Reasoning Capability in LLMs via Reinforcement Learning" (2025)
- HuggingFace TRL GRPOTrainer integration docs
- Lambert et al., "Tulu 3" (post-training methods survey, 2024)

## Machine

po-2024 (RTX 3070 8 Go) pioneering. Approfondissement possible sur ai-01 (gros GPU) si exploration full FT vs LoRA.

## Priorite

**LIVRABLE CLE Epic #1742**. Priorite haute apres PT-03 DPO. Sub-issue #1742-D.

**Comment by jsboige** [comment]
Delivered via PR #1766 (#1756), #1768 (#1757), #1771 (#1758), #1772 (#1759). All merged to main. Epic #1742 COMPLETE.
